### PR TITLE
Implement error for error types

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Builds on top of https://github.com/Concordium/concordium-contracts-common/pull/46.

Closes https://github.com/Concordium/concordium-contracts-common/issues/43.

## Changes

- Implement `std::error::Error` when the `std` feature is enabled

When compared to https://github.com/Concordium/concordium-contracts-common/pull/46: none of the examples change in size.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-contracts-common/src/types.rs [hunk 1]
```diff
@@ -78,6 +78,9 @@ impl fmt::Display for AmountParseError {
     }
 }
 
+#[cfg(feature = "std")]
+impl std::error::Error for AmountParseError {}
```
Comment (abizjak) [inline concordium-contracts-common/src/types.rs:82]: So this is a bit wonky. It uses the default implementation which will have `.description` return "Deprecated, use display".

Is this intended?

File: concordium-contracts-common/src/types.rs [hunk 2]
```diff
@@ -393,6 +396,9 @@ impl fmt::Display for ParseTimestampError {
     }
 }
 
+#[cfg(feature = "derive-serde")]
+impl std::error::Error for ParseTimestampError {}
```
Comment (abizjak) [inline concordium-contracts-common/src/types.rs:400]: Likewise here.

File: concordium-contracts-common/src/types.rs [hunk 3]
```diff
@@ -1310,6 +1325,13 @@ pub struct ParseError {}
 /// of parsing from binary via the `Serial` instance.
 pub type ParseResult<A> = Result<A, ParseError>;
 
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { write!(f, "Parsing failed") }
```
Comment (abizjak) [inline concordium-contracts-common/src/types.rs:1329]: Maybe better to use the more precise `f.write_str` in this case.

**Review by abizjak — COMMENTED** [review]
Could we maybe use `thiserror` in this crate to derive Display + Error types?

File: concordium-contracts-common/src/types.rs (hunk 1 above)
Comment (limemloh) [reply concordium-contracts-common/src/types.rs:82]: That is the current intended way of implementing this, and seems to also be how the `thiserror::Error` derive macro behaves.

**Review by abizjak — APPROVED** [review]

**Pull request merged**